### PR TITLE
- Implement registration sharing

### PR DESCRIPTION
**Pull request by rjschwei**
+ Registration sharing allows two completely independently configured
    SMT servers to be configured as sibling servers. The sibling servers
    share registration information effectively creating an redundant setup.
    Code that can be implemented in the client can detect if the SMT server
    that it registered to is available, if not it can fail over to the sibling
    server and get access to the repositories with the same registration
    credentials.

### CODE REVIEW

File: www/perl-lib/SMT/NCCRegTools.pm [hunk 1]
```diff
@@ -103,8 +103,30 @@ sub new
     $self->{URI}      = $ruri;
     $self->{AUTHUSER} = $user;
     $self->{AUTHPASS} = $pass;
-    bless($self);
 
+    my $regsharing = SMT::Utils::hasRegSharing();
```
Comment (mcalmer) [inline www/perl-lib/SMT/NCCRegTools.pm]: I do not like that "hasRegSharing" returns a string true/false which both is true if you check the value in the if statement.
If you read the code you have the feeling that you always get an error if regsharing is disabled.
See also comment on Utils.pm

File: www/perl-lib/SMT/Utils.pm [hunk 2]
```diff
@@ -37,6 +37,8 @@ use English;
 our @ISA = qw(Exporter);
 our @EXPORT = qw(__ __N printLog LOG_ERROR LOG_WARN LOG_INFO1 LOG_INFO2 LOG_DEBUG LOG_DEBUG2 LOG_DEBUG3);
 
+my $REGSHARING;
+
```
Comment (mcalmer) [inline www/perl-lib/SMT/Utils.pm]: Define $REGSHARING to 0 (false)

File: www/perl-lib/SMT/Utils.pm [hunk 3]
```diff
@@ -127,6 +129,47 @@ sub db_connect
...
+            {
+                return;
+            }
+            return http_fail($r, 500, $msg);
+        }
+        my $allowedSenders = $cfg->val('LOCAL', 'acceptRegistrationSharing');
+        my $shareRegDataTargets = $cfg->val('LOCAL', 'shareRegistrations');
+
+        if ($allowedSenders && $shareRegDataTargets) {
+            $REGSHARING = 'true';
```
Comment (mcalmer) [inline www/perl-lib/SMT/Utils.pm]:
```
$REGSHARING = 1;
```
and remove the "else"

File: www/perl-lib/SMT/Utils.pm [hunk 4]
```diff
@@ -127,6 +129,47 @@ sub db_connect
...
+        } else {
+            $REGSHARING = 'false';
+        }
+    }
+
+    return $REGSHARING;
+}
+
+
 =item __()
```
Comment (mcalmer) [inline www/perl-lib/SMT/Utils.pm]: If you want to read the values only one time, you can set 
```
$REGSHARING = undef;
```

and say:

```
if (! defined $REGSHARING) {
    $REGSHARING = 0;
    eval {
...
```

File: www/perl-lib/SMT/Registration.pm [hunk 5]
```diff
@@ -51,6 +53,28 @@ sub handler {
         $r->log_error("protocol version '".$hargs->{'version'}."' not implemented");
         return SMT::Utils::http_fail($r, 400, "Invalid protocol version.");
     }
+
+    # check if registration sharing is enabled and load module if it is
+    if (! $REGSHARING) {
+        $REGSHARING = SMT::Utils::hasRegSharing($r);
```
Comment (mcalmer) [inline www/perl-lib/SMT/Registration.pm]: If hasRegSharing returns 0/1 you can call the function directly everywhere where you need it.
If this function caches the value, you do not need to cache it again.

File: www/perl-lib/SMT/Rest/SCCAPIv1.pm [hunk 6]
```diff
@@ -10,14 +10,30 @@ use JSON;
...
+        };
+        if ($@)
+        {
+            my $msg = 'Failed to load registration sharing module '
+                . '"SMT/RegistrationSharing.pm"'
+                . "\n$@";
+            $r->log_error($msg);
+            $self->{regsharing} = 'false';
+        }
+    }
```
Comment (mcalmer) [inline www/perl-lib/SMT/Rest/SCCAPIv1.pm]: I think this block is something for Base.pm. 
Base is the right class for initializations like this.

**Comment by rjschwei** [comment]
OK, I think I have it sorted out now. As far as "caching" in multiple places is concerned I don't really see a way around that as we have 2 conditions, 1 - is reg sharing configured, 2- did the module load. In places where the reg sharing is used the cached value indicates the 2nd condition, in the utils it indicates the first.

File: www/perl-lib/SMT/Utils.pm [hunk 7]
```diff
@@ -127,6 +129,46 @@ sub db_connect
...
+        my $cfg;
+        eval
+        {
+            $cfg = SMT::Utils::getSMTConfig();
+        };
+        if($@ || !defined $cfg)
+        {
+            if (! $r)
+            {
+                return -1;
```
Comment (mcalmer) [inline www/perl-lib/SMT/Utils.pm]: This "return -1"  will not work, because "-1" is true if check for a bool.
Return undef or 0 to indicate a "false"

File: www/perl-lib/SMT/NCCRegTools.pm [hunk 8]
```diff
@@ -103,8 +103,30 @@ sub new
...
+    if (! $regsharing) {
+        my $msg = 'Could not read SMT configuration file';
+        printLog($self->{LOG}, $self->vblevel(), LOG_WARN, __($msg));
+        $regsharing = 'false';
+    }
+    elsif ($regsharing eq 'true') {
```
Comment (mcalmer) [inline www/perl-lib/SMT/NCCRegTools.pm]: hasRegSharing() does not return the string 'true'. So this will not work. simply use only "else"

File: www/perl-lib/SMT/NCCRegTools.pm [hunk 9]
```diff
@@ -103,8 +103,30 @@ sub new
...
+        if ($@)
+        {
+            my $msg = 'Failed to load registration sharing module '
+              . '"SMT/RegistrationSharing.pm"'
+              . "\n$@";
+            printLog($self->{LOG}, $self->vblevel(), LOG_ERROR, __($msg));
+            $regsharing = 'false';
+        }
+    }
+    $self->{REGSHARING} = $regsharing;
```
Comment (mcalmer) [inline www/perl-lib/SMT/NCCRegTools.pm]: You need to re-think the assignment of this variable.
It has now 1 or the string 'false'.
Make it 0/1 everywhere and change the checks also at other places of this file.

File: www/perl-lib/SMT/Rest/SCCAPIv1.pm [hunk 10]
```diff
@@ -10,8 +10,7 @@ use JSON;
 
 use SMT::Utils;
 use SMT::Client;
-use SMT::Registration;
-
+use SMT;
```
Comment (mcalmer) [inline www/perl-lib/SMT/Rest/SCCAPIv1.pm]: I havn't tested this yet, but I do not see a reason for changing the import and I fear that now calling
SMT::Registration::findCatalogs() cannot be executed anymore.

Was this change on purpose or by accident?

File: www/perl-lib/SMT/Utils.pm [hunk 11]
```diff
@@ -127,6 +129,46 @@ sub db_connect
...
+        if($@ || !defined $cfg)
+        {
+            if (! $r)
+            {
+                return;
+            }
+            $r->log_error("Cannot read the SMT configuration file: ".$@);
+            my $msg = 'SMT server is missconfigured. Please contact your '
+                . 'administrator.';
+            return http_fail($r, 500, $msg);
```
Comment (mcalmer) [inline www/perl-lib/SMT/Utils.pm]: the return value of http_fail is an Apache2::Const::OK; which evaluate to "true". This would
indicate, that the loading of the module was successfull.
I would suggest to re-write this to:

```
  if ($r)
  {
      $r->log_error("Cannot read the SMT configuration file: ".$@);
  }
  return;
```

Comment (rjschwei) [inline www/perl-lib/SMT/Utils.pm]: Hmm, I was following the established pattern, see for example: Registration.pm lines 166-171

I have no issue with the code change, but would like to understand why this case would be different than other cases where we check for a successful load of the configuration.

File: www/perl-lib/SMT/Rest/SCCAPIv1.pm (hunk 10 above)
Comment (rjschwei) [inline www/perl-lib/SMT/Rest/SCCAPIv1.pm]: Oops, brain fart, fixed

File: www/perl-lib/SMT/Utils.pm (hunk 11 above)
Comment (mcalmer) [inline www/perl-lib/SMT/Utils.pm]: Registration.pm lines 166-171 is inside of the function register().
The result of register() is not checked.

Here in Utils.pm you write a function which should return a (kind of) boolean value - True, False and "undef" on error condition.
But http_fail return a true value and would lead into funny behavior.

I think you could test it easily by force some parsing errors or something like this.

Comment (rjschwei) [inline www/perl-lib/SMT/Utils.pm]: OK, got it now. Thanks. For some reason I thought http_fail would basically set the status and then exit the current execution path. Obviously it has been a while since I looked at the implementation of http_fail.

Anyway, in order to set the proper status (500) I elected to stick with using http_fail, but I modified the logic.

**Comment by mcalmer** [comment]
Merged into master
